src/modules/lua_mtev.c

```diff
@@ -953,9 +961,38 @@ static int
 mtev_lua_socket_sock_name(lua_State *L) {
   return mtev_lua_socket_gen_name(L, getsockname);
 }
+
+static int
+socket_connect_timeout(eventer_t e, int mask, void *closure, struct timeval *now) {
+  (void)mask;
+  (void)now;
+  mtev_lua_resume_info_t *ci;
+  lua_timeout_callback_ref* cb_ref;
+  lua_State *L;
+
+  cb_ref = (lua_timeout_callback_ref*) closure;
+  L = cb_ref->L;
+  ci = mtev_lua_find_resume_info(L, mtev_true);
+  assert(ci);
+
+  // remove the original event
+  eventer_remove_fde(cb_ref->timed_out_eventer);
+  mtev_lua_deregister_event(ci, cb_ref->timed_out_eventer, 0);
+
+  // return into the original Lua call which spawned this timeout
+  lua_pushinteger(L, -1);
+  lua_pushstring(L, "timeout");
+  ci->lmc->resume(ci, 2);
```

Comment:
socket:connect() returns `-1, "timeout"` on timeout. socket:read() returns nil. Hence the need for a separate function.